Feature/plot2dplotter

mafrahm opened the pull request:
This PR restructures the plotting tasks as shown in the comment below.
The major changes are that the `PlotVariables` task is now separated in two callable tasks,  `PlotVariables1d` and `PlotVariables2d`, which have the same base class but inherit from different base classes to provide different sets of parameters.

The new plotting function is implemented here [1] and uses all plotting parameters enabled for 2d plotting tasks except the `process-settings`.

Example command for `PlotVariables2d`:
```
law run cf.PlotVariables2d --version v1 --processes tt_sl,st_tchannel --variables n_jet_vs_jet1_pt --zscale log --shape-norm True
```
`PlotCutflowVariables2d` is also implemented and working, producing one plot per process and selector-step.


TODOs (for a next PR?):
- we should implement a `per-process` parameter for the 2d tasks to enable switching between producing 1 plot per process or 1 plot with all processes added together. This could also be a part of the `ProcessPlotSettingMixin`.
- use the process_settings in `plot_2d`. Changing plotting label and process scale could be used.
- The plot_function_names in `PlotCutflowVariables1d` are still hard-coded.

[1]

riga commented on columnflow/tasks/cutflow.py:
```diff
@@ -190,7 +200,7 @@ def get_plot_parameters(self):
 
         # no ratio plot implemented: disable ratio plot for `plot_cutflow`
-        if self.plot_function_name == "columnflow.plotting.example.plot_cutflow":
+        if self.plot_function_name_1d == "columnflow.plotting.example.plot_cutflow":
```
Maybe `skip_ratio` should rather become a parameter? Having plot function dependent settings somehow works around the function plug-in mechanism and might not scale too well.

riga commented on columnflow/util.py:
```diff
@@ -530,6 +530,16 @@ def __getattr__(self, attr: str) -> Any:
     def __setattr__(self, attr: str, value: Any) -> None:
         self[attr] = value
 
+    def cautious_update(self, other: dict) -> None:
+        """
+        Updates DotDict, but only if no value is overwritten. Throws error otherwise.
+        """
+        key_replicas = set(self.keys()) & set(other.keys())
+        if not key_replicas:
+            self.update(other)
+        else:
+            raise KeyError(f"'{self.__class__.__name__}' object already has attributes {key_replicas}")
+
```
2 things here:

- For the DotDict I think it would be better to keep the amount of "real" attributes as minimal as possible and thus rather move such a functionality into a util function that accepts any dict subtypes.
- Maybe more importantly: I think this way of doing "cautious" updates can be dangerous. IIUC, if any of the keys is already existing in the dict, then none of the other dict items are added. But then, one always needs to know whether certain keys are definitely not present, which is quite implicit and might depend on the situation. As an alternative, I would explicitly check for the presence of a certain key (at the point where the update should be performed), and then add new items depending on the response (see e.g. above)

dsavoiu replied:
I think the idea here was to have a convenience function that works as a regular `dict.update`, but have it fail if any of the keys to be added is already present in the dictionary. The main benefit of this is to prevent accidentally overwriting a value that was set by a previous function operating on the same `DotDict` (e.g. to prevent style changes from getting lost/clobbered when multiple plot functions are called successively). I'm not sure I would consider it dangerous, it would just fail instead of overwriting, so in a sense it functions as an assertion.

This could also be a util function instead of a member of `DotDict`, or one could avoid a function completely and check the presence of a key explicitly whenever a key is added, but having a function would avoid make the calling code less verbose.

It's true that the **values** are not checked. One could of course extend the implementation to fail not just if a key is already present, but only if the value for that key would also change through the update.

mafrahm replied:
I added the value check to the new `dict_add_strict` function :+1:

riga commented on columnflow/tasks/framework/plotting.py:
```diff
@@ -178,21 +183,22 @@ class PlotBase1D(PlotBase):
-        params["yscale"] = None if self.yscale == law.NO_STR else self.yscale
-        params["shape_norm"] = self.shape_norm
-
+        params_to_add = {
+            "skip_ratio": self.skip_ratio,
+            "yscale": None if self.yscale == law.NO_STR else self.yscale,
+            "shape_norm": self.shape_norm,
+        }
+        params.cautious_update(params_to_add)
```
See the last review comment first :)

The `cautious_update` can be dangerous here in case (e.g.) `yscale` is, say, already present and set to `None`. This would be fully consistent (since `None` would also be added through `params_to_add`), but still, `skip_ratio` and `shape_norm` would not be added to `params`.

Instead, maybe this is what one would like to read here as it is more explicit:

```python
... removed ...
```

... actually, I don't see a dependence among the three `params_to_add`, so I'm not sure why one would want to add all of them, but only if none of them is already present 🤔

dsavoiu replied:
To handle the case where (part of) a `cautious_update` is consistent with the existing values one could explicitly check if these change (see above).

riga replied:
Indeed. With the strict add method, this should be fine.

riga reviewed (changes requested):
Nice work 🎉 I'm eager to see the 2D plot style.

However, I couple comments inline, mainly related to the `cautious_update` which I find too implicit and therefore dangerous.

riga commented on columnflow/tasks/cutflow.py:
```diff
@@ -190,7 +200,7 @@ def get_plot_parameters(self):
 
         # no ratio plot implemented: disable ratio plot for `plot_cutflow`
-        if self.plot_function_name == "columnflow.plotting.example.plot_cutflow":
+        if self.plot_function_name_1d == "columnflow.plotting.example.plot_cutflow":
```
Maybe `skip_ratio` should rather become a parameter? Having plot function dependent settings somehow works around the function plug-in mechanism and might not scale too well.

mafrahm replied:
`skip_ratio` already is a parameter, I just set it to True here, because this plot function does not provide a ratio plot. But I think, the ratio plot would not be produced event when `skip_ratio=False`, so I could also just remove this line.

riga replied:
👍

mafrahm replied:
turns out I can't just remove this because then an empty ratio plot is drawn, but I simply set `skip_ratio=True` in the plot_function itself

mafrahm commented:
There were still some issues with handling of line colors of stacks. I fixed this with another commit, where I also moved some duplicated code into this file:
https://github.com/uhh-cms/columnflow/blob/feature/plot2dplotter/columnflow/plotting/plot_util.py

dsavoiu reviewed (approved):
Very nice, this is definitely useful! I had a few comments about the `cautious_update` function, but it seems like `dict_add_strict` fits in quite nicely as an alternative. Looks good to me otherwise, so I'm fine with merging these changes.

riga reviewed (changes requested):
Just some minor left-over comment.

mafrahm commented on columnflow/tasks/framework/plotting.py:
```diff
@@ -257,23 +259,33 @@ def store_parts(self):
     Base class for plotting tasks creating 2-dimensional plots.
     """
 
+    plot_function_name_2d = luigi.Parameter(
```
I agree, this change is pushed
